File: 2023/C/src/day1.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include "inc/days.h"

int parseByLine(FILE *fp);
int firstLast(const char *str);
int getDigit(const char *str);
int textToDigit(const char *buf);
int startsWith(const char *pre, const char *str);

int runDay1(const char *file) {
  FILE *fp = fopen(file, "r");
  if (fp == NULL) {
    fprintf(stderr, "ERR: Can't open file '%s'\n", file);
    exit(1);
  }

  int out = parseByLine(fp);

  fclose(fp);
  return out;
}

int parseByLine(FILE *fp) {
  //int lines = 0;
  int out = 0;
  char buf[256];

  while (fgets(buf, 256, fp) != NULL) {
    int fl = firstLast(buf);
    out += fl;
    //printf("%4d: [%2d] %s", lines, fl, buf);
    //lines++;
  }

  return out;
}

int firstLast(const char *str) {
  if (str == NULL) { return 0; }
  unsigned long len = strlen(str);
  if (len <= 0) { return 0;}
  len = strlen(str) - 1;
  int s = 0;
  int e = 0;

  for (int i = 0; i < len; i++) {
    if (s != 0 && e != 0) { break; }
    if (s == 0) { s = getDigit(&str[i]); }
    if (e == 0) { e = getDigit(&str[len - (i + 1)]); }
  }

  return (s * 10) + e;
}

int getDigit(const char *str) {
  if (isdigit(str[0]) != 0) {
    return str[0] - '0';
  } 

  return textToDigit(str);
}

int textToDigit(const char *buf) {
  if (startsWith("one", buf) == 0) { return 1; }
  if (startsWith("two", buf) == 0) { return 2; }
  if (startsWith("three", buf) == 0) { return 3; }
  if (startsWith("four", buf) == 0) { return 4; }
  if (startsWith("five", buf) == 0) { return 5; }
  if (startsWith("six", buf) == 0) { return 6; }
  if (startsWith("seven", buf) == 0) { return 7; }
  if (startsWith("eight", buf) == 0) { return 8; }
  if (startsWith("nine", buf) == 0) { return 9; }
  return 0;
}

int startsWith(const char *pre, const char *str) {
  return strncmp(pre, str, strlen(pre));
}

File: 2023/JS/day1.mjs
'use strict'
import * as fs from 'node:fs';

const english_digit = {
  'one': 1,
  'two': 2,
  'three': 3,
  'four': 4,
  'five': 5,
  'six': 6,
  'seven': 7,
  'eight': 8,
  'nine': 9
};

function isDigit(c) {
  return c >= '0' && c<='9';
}

function getAsDigit(str, offset) {
  if (isDigit(str[offset])) {
    return str[offset] - '0';
  }

  for (const [key, value] of Object.entries(english_digit)) {
    if (str.startsWith(key, offset)) { return value; }
  }

  return null;
}

function day1(file) {
  let data;

  try {
    data = fs.readFileSync(file, 'utf8');
  } catch (err) {
    console.log(`ERR: ${err}`);
    return;
  }
  
  let sum = 0;
  const lines = data.split('\n');
  lines.forEach(line => {
    let start = 0;
    let end = 0;
    for (let i = 0; i <= line.length; i++) {
      if (start != 0 && end != 0) {
        let combined = start * 10 + end;
        sum += combined;
        break;
      }

      if (start === 0) {
        let start_digit = getAsDigit(line, i);
        if (start_digit !== null) {
          start = start_digit - '0';
        }
      }

      if (end === 0) {
        let end_digit = getAsDigit(line, line.length - (i + 1));
        if (end_digit !== null) {
          end = end_digit - '0';
        }
      }
    }
  });

  return sum;
}

console.log('DAY 1');
console.log('-------------------------------------------------------------');
const out = day1('data/official_input_day1');
console.log(`OUT: ${out2}`);
